test/widgets/test_generic_poll_text.py

```diff
@@ -0,0 +1,118 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Widget specific tests
+
+import contextlib
+import sys
+from importlib import reload
+from types import ModuleType
+
+import pytest
+
+from libqtile.widget import generic_poll_text
+
+
+class Mockxml(ModuleType):
+    @classmethod
+    def parse(cls, value):
+        return {"test": value}
+
+
+class MockRequest:
+    return_value = None
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+
+class Mockurlopen:
+    def __init__(self, request):
+        self.request = request
+
+    class headers:
+        @classmethod
+        def get_content_charset(cls):
+            return "utf-8"
+
+    def read(self):
+        return self.request.return_value
+
+
+@contextlib.contextmanager
+def inject_mock_module(classdef, name, return_value=None):
+    sys.modules[name] = classdef(name)
+    yield None
+    del sys.modules[name]
+
+
+def test_gen_poll_text():
+    gpt_no_func = generic_poll_text.GenPollText()
+    assert gpt_no_func.poll() == "You need a poll function"
+
+    gpt_with_func = generic_poll_text.GenPollText(func=lambda: "Has function")
+    assert gpt_with_func.poll() == "Has function"
+
+
+def test_gen_poll_url_not_configured():
+    gpurl = generic_poll_text.GenPollUrl()
+    assert gpurl.poll() == "Invalid config"
+
+
+def test_gen_poll_url_text(monkeypatch):
+    gpurl = generic_poll_text.GenPollUrl(json=False, parse=lambda x: x, url="testing")
+    monkeypatch.setattr(generic_poll_text, "Request", MockRequest)
+    monkeypatch.setattr(generic_poll_text, "urlopen", Mockurlopen)
+    generic_poll_text.Request.return_value = b"OK"
+    assert gpurl.poll() == "OK"
+
+
+def test_gen_poll_url_json(monkeypatch):
+    gpurl = generic_poll_text.GenPollUrl(parse=lambda x: x, data=[1, 2, 3], url="testing")
+    monkeypatch.setattr(generic_poll_text, "Request", MockRequest)
+    monkeypatch.setattr(generic_poll_text, "urlopen", Mockurlopen)
+    generic_poll_text.Request.return_value = b'{"test": "OK"}'
+    assert gpurl.poll()["test"] == "OK"
+
+
+def test_gen_poll_url_xml(monkeypatch):
+    gpurl = generic_poll_text.GenPollUrl(json=False, xml=True, parse=lambda x: x, url="testing")
+    monkeypatch.setattr(generic_poll_text, "Request", MockRequest)
+    monkeypatch.setattr(generic_poll_text, "urlopen", Mockurlopen)
+    generic_poll_text.Request.return_value = b"OK"
+    with pytest.raises(Exception):
+        gpurl.poll()
+
+    with inject_mock_module(Mockxml, "xmltodict"):
+        reload(generic_poll_text)
```

Comment:
I've inkected my fake `xmltodict` module but I have to reload the widget module as the `ImportError` test is only run once when the module is loaded.

Reply:
Can you make this a comment in the code?